import "maidsafe-dht/kademlia/kademlia.proto";
import "maidsafe-dht/transport/transport.proto";

package maidsafe.kademlia.protobuf;

message SignedValue {
  required bytes value = 1;
  required bytes signature = 2;
}

message PingRequest {
  required Contact sender = 1;
  required bytes ping = 2;
  optional bytes public_key = 3;
}

message PingResponse {
  required bytes echo = 1;
}

message FindValueRequest {
  required Contact sender = 1;
  required bytes key = 2;
  optional bytes public_key = 3;
}

message FindValueResponse {
  required bool result = 1;
  repeated Contact closest_nodes = 2;
  repeated SignedValue signed_values = 3;
  optional Contact alternative_value_holder = 4;
}

message FindNodesRequest {
  required Contact sender = 1;
  required bytes key = 2;
  optional bytes public_key = 3;
}

message FindNodesResponse {
  required bool result = 1;
  repeated Contact closest_nodes = 2;
}

message StoreRequest {
  required Contact sender = 1;
  required bytes key = 2;
  required SignedValue signed_value = 3;
  required int32 ttl = 4;
  optional bytes public_key = 5;
  optional bytes serialised_store_request = 6;
  optional bytes serialised_store_request_signature = 7;
}

message StoreResponse {
  required bool result = 1;
}

message DeleteRequest {
  required Contact sender = 1;
  required bytes key = 2;
  required SignedValue signed_value = 3;
  optional bytes public_key = 4;
  optional bytes serialised_delete_request = 5;
  optional bytes serialised_delete_request_signature = 6;
}

message DeleteResponse {
  required bool result = 1;
}

message DownlistNotification {
  required Contact sender = 1;
  repeated bytes node_ids = 2;
}
